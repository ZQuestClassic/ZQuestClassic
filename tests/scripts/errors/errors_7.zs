#option NO_ERROR_HALT on

void test()
{
	int arr[] = {1,2,3};
	int[] arr2[] = {arr,arr};
	npc[] enemies = Screen->NPCs;

	// Due to legacy array type conversion, the default here is to warn.
	arr2[0] = 5;
	arr[0] = arr2;
	arr2[2] = arr2;
	int[] arr3_bad = {arr,arr};
	{
		#option LEGACY_ARRAYS off

		// Error.
		arr2[0] = 5;
		arr[0] = arr2;
		arr2[2] = arr2;
		int[] arr3_bad = {arr,arr};
	}
	{
		#option LEGACY_ARRAYS on

		// OK.
		arr2[0] = 5;
		arr[0] = arr2;
		arr2[2] = arr2;
		int[] arr3_bad = {arr,arr};
	}

	//Errors
	arr2[1] = enemies;
	arr[0] = enemies[0];
	arr2[1][0] = enemies[1];

	//OK
	arr[0] = <int>arr2;
	untyped foo = arr;
	foo = arr2;
	foo = enemies;

	//!TODO
	arr2[1] = {2,3,4}; //!TODO should be valid, if arrays become gc-managed
}

void arraycat<T>(T[] dest, T[] vals)
{
	int indx = SizeOfArray(dest);
	ResizeArray(dest, indx+SizeOfArray(vals));
	for(v : vals)
		dest[indx++] = v;
}

void append<T>(T[] arr, T val)
{
	ArrayPushBack(arr,val);
}

void print<T>(T v)
{
	printf("Val: %d\n", v);
}
void print<T>(T[] arr)
{
	printf("Arr[]: %ad\n", arr);
}
void print<T>(T[][] arr)
{
	printf("Arr[][]: %aad\n", arr);
}
