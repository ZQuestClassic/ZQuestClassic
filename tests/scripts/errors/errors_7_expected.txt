Compiling 'errors_7.zs'
Pass 1: Parsing
Pass 2: Preprocessing
Pass 3: Registration
Pass 4: Analyzing Code
ZQ_BUFFER Line 9 @ Columns 2-13 - Error T017: Cannot cast from int to int[].
ZQ_BUFFER Line 10 @ Columns 2-15 - Error T017: Cannot cast from int[][] to int.
ZQ_BUFFER Line 11 @ Columns 2-16 - Error T017: Cannot cast from int[][] to int[].
ZQ_BUFFER Line 12 @ Columns 2-19 - Error T017: Cannot cast from npc[] to int[].
ZQ_BUFFER Line 13 @ Columns 2-21 - Error T017: Cannot cast from npc to int.
ZQ_BUFFER Line 14 @ Columns 2-25 - Error T017: Cannot cast from npc to int.
ZQ_BUFFER Line 23 @ Columns 12-19 - Error T100: Temporary literals cannot be stored!
Failure!
Compile finished with exit code '17' (compiled with errors)
